src/Elders.Cronus/Hosting/CronusServiceCollectionExtensions.cs

```diff
@@ -14,28 +14,50 @@ namespace Elders.Cronus
     public static class CronusServiceCollectionExtensions
     {
         /// <summary>
-        /// // Adds Cronus core services
+        /// Adds Cronus core services
         /// </summary>
         public static IServiceCollection AddCronus(this IServiceCollection services, IConfiguration configuration)
         {
-            return AddCronus(services, new CronusServicesProvider(services, configuration));
+            return services.AddCronus(configuration, new CronusServicesDiscoveryHandler(services, configuration));
         }
 
         /// <summary>
-        /// // Adds Cronus core services
+        /// Adds Cronus core services
         /// </summary>
-        public static IServiceCollection AddCronus(this IServiceCollection services, CronusServicesProvider cronusServicesProvider)
+        public static IServiceCollection AddCronus(this IServiceCollection services, IConfiguration configuration, ICronusServicesDiscoveryHandler cronusServicesProvider)
         {
+            services.AddSingleton(configuration);
+            services.AddSingleton(cronusServicesProvider);
+            services.AddDiscoveryScanner();
             services.AddTenantSupport();
             services.AddCronusHostOptions();
             services.AddDefaultSubscribers();
 
-            var discoveryFinder = new DiscoveryScanner();
-            var discoveryContext = new DiscoveryContext(AssemblyLoader.Assemblies.Values, cronusServicesProvider.Configuration);
-            var discoveryResults = discoveryFinder.Scan(discoveryContext);
+            return services;
+        }
+
+        /// <summary>
+        /// Builds a <see cref="ServiceProvider"/> for the discovered Cronus services
+        /// </summary>
+        public static ServiceProvider BuildCronusServiceProvider(this IServiceCollection services)
+        {
+            var serviceProvider = services.BuildServiceProvider();
+            CronusLogger.Bootstrap(serviceProvider);
+            var configuration = serviceProvider.GetRequiredService<IConfiguration>();
+            var discoveryScanner = serviceProvider.GetRequiredService<DiscoveryScanner>();
+            var discoveryContext = new DiscoveryContext(AssemblyLoader.Assemblies.Values, configuration);
+            var discoveryResults = discoveryScanner.Scan(discoveryContext);
 
+            var cronusServicesDiscoveryHandler = serviceProvider.GetRequiredService<ICronusServicesDiscoveryHandler>();
             foreach (var result in discoveryResults)
-                cronusServicesProvider.HandleDiscoveredModel(result);
+                cronusServicesDiscoveryHandler.HandleDiscoveredModel(result);
+
+            return cronusServicesDiscoveryHandler.Services.BuildServiceProvider();
+        }
+
+        internal static IServiceCollection AddDiscoveryScanner(this IServiceCollection services)
+        {
+            services.AddSingleton<DiscoveryScanner>();
```

Comment:
The discovery scanner is executed before building the ServiceProvider. It makes no sense to add it in the services collection.